File: pairSum.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
import java.util.Stack;
class Solution {
    public int pairSum(ListNode head) {
        Stack<Integer> twins = new Stack<>();
        ListNode iter = head;
        int size=0;
        while(iter!=null) {
            twins.push(iter.val);
            iter=iter.next;
            size+=1;
        }
        iter=head;
        int res=Integer.MIN_VALUE;
        int twin=Integer.MIN_VALUE;
        int count=0;
        while(!twins.isEmpty()){
            twin=twins.pop();
            res=Math.max(res,twin+iter.val);
            iter=iter.next;
            count+=1;
            if (count==size/2) break;
        }
        return res;
        
    }
}

File: pairSum.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        slow=head
        fast=head
        res=-sys.maxsize
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next
        prev=None
        curr=head
        while(curr.next!=slow):
            next=curr.next
            curr.next=prev
            prev=curr
            curr=next
        curr.next=prev
        while(slow and curr):
            res=max(res,slow.val+curr.val)
            slow=slow.next
            curr=curr.next

        return res
        
        
